docs: One-pager for node operators (#3468)
near/mpc: docs/operating-an-mpc-node.md

# MPC operator expectations

> Whether you're already running a node or considering it — thank you. The NEAR MPC network stays secure, reliable, and decentralized only because independent operators like you keep it running. No single party can produce a signature alone; that guarantee rests on you.

Reviewer: Hann that's sweet

Author: 😊

This guide lays out what we expect from operators. If anything is unclear, raise it in the operators' Slack channel.

## What you run
All operators are expected to run the following:

1. A TEE testnet MPC node

> 2. A TEE mainnet MPC node

Reviewer: Nit: I would make testnet and mainnet in bold.

Author: Hmm in this case I don't see how it's worth emphasizing testnet and mainnet. Especially since they should run both. Can you help me understand how that would be helpful for the reader?

Reviewer: When I was reading that I bugged for a couple seconds on why have written the same sentence twice. But it wasn't

Author: Ah I see, sure I can make them bold.

Author: Updated in https://github.com/near/mpc/pull/3468/commits/e16485d8790767bc56798b64c1b093472e9250b0

> 3. (once implemented) Our TEE enabled backup service

Reviewer: Any links for those steps?

Author: Added https://github.com/near/mpc/pull/3468/commits/9a39a952ee460b881e2afd7a8e8bd1ca2f64cc0e

> 3. (once implemented) Our TEE enabled backup service

Author: Sure, makes sense.

> Do not co-locate these on the same machine — each needs its own. We also want to keep our testnet and mainnet setups as identical as possible to minimize the risk of mainnet-only bugs.

Author: There's a difference between what's possible and what we recommend for operators of our MPC network. But yeah, this discrepancy is not ideal. However, I prefer the existing formulation for this doc. For the operators of the live network, this is an expectation - not a recommendation.

> Do not co-locate these on the same machine — each needs its own. We also want to keep our testnet and mainnet setups as identical as possible to minimize the risk of mainnet-only bugs.

Reviewer 2: I would make these two separate points:
- each component its own machine
- try to keep mainnet and testnet setups as close to each other as possible.

Author: Agree, done in https://github.com/near/mpc/pull/3468/commits/10a1272929ebeaadd252f78172bd50187571a674

See the [TDX node guide](./running-an-mpc-node-in-tdx-external-guide.md) for how to operate TEE MPC nodes, and the [node migration guide](./node-migration-guide.md) for how to use our backup CLI to migrate nodes between different environments.

> > **Note** — TEE migration is in progress: testnet runs a mix of TEE and non-TEE nodes today, and mainnet has no TEE nodes yet. Until that changes, operators onboarding now should run a non-TEE node for mainnet and a TEE node for testnet. See the [legacy setup guide](https://docs.google.com/document/d/1YQFgjPxqP3U0w9wRxhZsTNQYyUjsD_y3jAwzNYhyNcM/) for details.

Reviewer: I would also go with Copilot's suggestion

> > **Note** — TEE migration is in progress: testnet runs a mix of TEE and non-TEE nodes today, and mainnet has no TEE nodes yet. Until that changes, operators onboarding now should run a non-TEE node for mainnet and a TEE node for testnet. See the [legacy setup guide](https://docs.google.com/document/d/1YQFgjPxqP3U0w9wRxhZsTNQYyUjsD_y3jAwzNYhyNcM/) for details.

Author: Yeah applied it

### One node per network per operator

> Each operator runs exactly one mainnet node and one testnet node — no more. A decentralized set of independently operated nodes is the whole point, so no single party may run more than one node per network, whether directly or by quietly operating several operators' nodes through subcontracting.

Author: This refers to MPC nodes, I think talking about the backup service here makes the paragraph more confusing. If operators are confused by this, I'd be happy to change.

> Each operator runs exactly one mainnet node and one testnet node — no more. A decentralized set of independently operated nodes is the whole point, so no single party may run more than one node per network, whether directly or by quietly operating several operators' nodes through subcontracting.

Reviewer: Very important paragraph. Can you somewhat highlight it? Can you talk about the penalization in case they subcontract?
Does it make sense to define what we mean by subcontracting?

Author: I think subcontracting is pretty well-defined: https://dictionary.cambridge.org/dictionary/english/subcontracting

> the act of paying an outside person or organization to do work that might normally be done within an organization

This should be uncommon but we did have a case where the same node operator was asked to run a node for another operator. Which is why we mention it here in the first place.

I also don't want to focus on penalization in this document. That's between the operators and House of Stake. The purpose of this doc is to clarify expectations on node operators. The SLAs define the formal requirements, which includes much more elaborate legal language and details about payouts.

Author: Though highlighting **exactly one** makes sense, I'll do it.

Author: Done in https://github.com/near/mpc/pull/3468/commits/9a39a952ee460b881e2afd7a8e8bd1ca2f64cc0e

> Subcontracting the day-to-day operation of *your* node is fine, but the same expectations apply to whoever ends up running it.

Reviewer 2: Hmm, is the point of this subsection to say that no two node operators must use the same subcontractor? So, lets say Alice and Bob both subcontract Charlie to run their node. We wouldn't allow this? How is this enforced? Do they need to inform HoS who they subcontract to?

Author: > Hmm, is the point of this subsection to say that no two node operators must use the same subcontractor? 

In the context of this doc it's not **must**, but rather **should**. This document outlines the expectations. The legal requirements are in the operator's SLAs.

> So, lets say Alice and Bob both subcontract Charlie to run their node. We wouldn't allow this? 

Yes, this would reduce decentralization so we don't want operators to do this. Or as in a specific case that we pushed back against - Alice is a node operator, and Bob wants to subcontract Alice to run Bob's node alongside Alice's node.

> We wouldn't allow this? How is this enforced? Do they need to inform HoS who they subcontract to?

Technically we can't enforce it, but it becomes very obvious if the same organization is the contact point for two separate nodes, and we wouldn't vote to add them to the operator set.

### Keep your operator account isolated
Use your MPC operator account only for running the node — don't call other smart contracts with it. This keeps it isolated and adds defense-in-depth against malicious contracts forwarding requests to the signer contract.

## Staying reachable

> We have a shared Slack channel for all node operators. We expect you to be available there to coordinate network upgrades and respond to incidents.

Reviewer 2: Should we add a link to the slack channel in this doc?

Author: No I don't think that link is important here. All operators know which channel there is, and for new operators the only relevant information is that there is a slack channel. The exact link doesn't help them get added.

## Uptime

> You're expected to keep your nodes healthy and participating in their networks. During normal operation we expect 99% uptime. That may sound low for this kind of service, but because the MPC network tolerates individual node failures, it translates to very high uptime for the network as a whole.

Reviewer: I feel 99% is quite high not low..

Author: Depends on what you compare it to. For example Amazon compute has 99.99% uptime in their SLA https://aws.amazon.com/compute/sla/

And for S3 data durability is 11 nines, see https://docs.aws.amazon.com/AmazonS3/latest/userguide/DataDurability.html

> Designed to provide 99.999999999% durability and 99.99% availability of objects over a given year.

just two nines implies ~7 hours of downtime each month which is below what's acceptable for most services imo.

Reviewer: Wow! Really interesting

### Excused downtime
You're not expected to run hot backup nodes for quick recovery if a node crashes for reasons outside your control. Syncing a recovering node can take a while — that's fine, and won't count against your uptime expectations as long as you responded promptly and took appropriate action.

> The same applies if we ask you to turn off your node during incident investigations or similar situations.

Reviewer: What kind of trouble shooting mechanism do we have? Does it make sense to note it here?

Author: I want to keep this document brief. Specifics about how we troubleshoot is the purpose of our on-call runbooks.

## Incident response

> We monitor network health. If something looks wrong with your node, we'll flag it to you on Slack, and we expect you to pick it up and start investigating as soon as you reasonably can. Prompt response is essential to the reliability of the network.

Reviewer: We cannot count only on slack. What if we have a problem while slack is down? We should have a second mean of communication that is automatic and instant. However slack is prioritized if both means are working

Author: Agree, the current situation is not ideal. We should add better alerting/paging mechanisms as highlighted in the next paragraph.

However, this is how it works now and this document reflects that.

> We monitor network health. If something looks wrong with your node, we'll flag it to you on Slack, and we expect you to pick it up and start investigating as soon as you reasonably can. Prompt response is essential to the reliability of the network.

Reviewer 2: Does that mean we take responsibility to flag issues for _each_ node operator?
Who is "we" in this?

Author: Yeah, "we" are Near one. Though all operators are expected to monitor the network, and the purpose of this section is to set expectations on responsiveness for incidents. I'll come up with a better formulation.

Author: I use "we" in a few other places, so I clarified in the beginning who we are in https://github.com/near/mpc/pull/3468/commits/e271f187eda3cb6551a4ed8a3206d7eea96f59bc

Also updated this sentence so it doesn't sound like we take responsibility for flagging issues.

> We'll move to tighter, formal response times once dedicated alerting is in place; until then, prompt best-effort response is what matters.

Author: I used to have this, but realized it doesn't make sense to explicitly state hard expectations as long as we don't have proper paging of operators set up.

When an incident is underway we expect you to be ready to collaborate on actions such as:

- Restarting your node.
- Updating your node.
- Submitting votes or other information to the smart contract.
- Sharing logs for debugging.

## Upgrades
When we publish new MPC versions we typically coordinate node upgrades first, followed by contract upgrades. We coordinate both on Slack. In each case we'd like you to independently verify the code before acting, so you can confirm the expected version is what's actually being run.

### Node upgrades
Follow our [reproducible builds guide](./reproducible-builds.md) to verify the docker image hashes, then upgrade your nodes to the requested version.

### Contract upgrades & voting
Contract upgrades happen through voting. Just like with nodes, reproduce the contract build to verify the code before voting.

> For any on-chain vote — upgrades, generating new keys, or changes to the operator set — we expect you to respond within three (3) calendar days of it opening: either approve it, or raise an objection so we can take it from there.

Reviewer 2: nit:
```suggestion
For any on-chain vote — upgrades, generating new keys, or changes to the operator set — we expect you to respond within three (3) calendar days of it opening: either approve it, or raise an objection on slack.
```

"take it from there"  does not add any value. It would be more helpful to explain how they can raise an objection.

Author: Good point, applied in https://github.com/near/mpc/pull/3468/commits/1fd6550d5a89a567c0963ec3f9c8a1d8a1e4b240

### One node per network per operator
Each operator runs **exactly one** mainnet node and one testnet node — no more. A decentralized set of independently operated nodes is the whole point, so no single party may run more than one node per network, whether directly or by quietly operating several operators' nodes through subcontracting.

Subcontracting the day-to-day operation of *your* node is fine, but the same expectations apply to whoever ends up running it.

### Keep your operator account isolated
Use your MPC operator account only for running the node — don't call other smart contracts with it. This keeps it isolated and adds defense-in-depth against malicious contracts forwarding requests to the signer contract.

## Staying reachable
We have a shared Slack channel for all node operators. We expect you to be available there to coordinate network upgrades and respond to incidents.

## Uptime

> You're expected to keep your nodes healthy and participating in their networks. During normal operation we expect 99% uptime. That may sound low for this kind of service, but because the MPC network tolerates individual node failures, it translates to very high uptime for the network as a whole.

Reviewer 3: ```suggestion
You're expected to keep your nodes healthy and participating in their networks. During normal operation we expect 99% uptime. That may sound low for this kind of service, but because the MPC network tolerates individual node failures without any user impact, it translates to very high uptime for the network as a whole.
```

Author: Thanks, applied in https://github.com/near/mpc/pull/3468/commits/c8ba678b0a37b92480a85693a1afc88cb08629e3

### Excused downtime
You're not expected to run hot backup nodes for quick recovery if a node crashes for reasons outside your control. Syncing a recovering node can take a while — that's fine, and won't count against your uptime expectations as long as you responded promptly and took appropriate action.

The same applies if we ask you to turn off your node during incident investigations or similar situations.

## Incident response
When an issue with your node comes up, we expect you to pick it up and start investigating as soon as you reasonably can, and to keep others posted on Slack as you work it. This is essential to the reliability of the network.

We'll move to tighter, formal response times once dedicated alerting is in place; until then, prompt best-effort response is what matters.

When an incident is underway we expect you to be ready to collaborate on actions such as:

- Restarting your node.
- Updating your node.
- Submitting votes or other information to the smart contract.
- Sharing logs for debugging.

> ## Upgrades

Reviewer 3: IMO we should mention that NEAR protocol upgrades always require an MPC upgrade as well. Also, can we mention a rough estimate of how often this all happens, e.g. an upgrade (due to the protocol upgrade or new MPC release) happens roughly every 2 weeks.

Author: Good point, updated the formulation in https://github.com/near/mpc/pull/3468/commits/87d7d82d668f34787d494d04e9e0e3e893fcf268

Author: Also fun fact, we've made 27 releases since 3.0.0 in November last year - averaging one release every 7.8 days. Though if we exclude patch releases it's on average 17 days between releases instead.
